syz-verifier/main.go

```diff
@@ -0,0 +1,310 @@
+// Copyright 2021 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+package main
+
+import (
+	"flag"
+	"log"
+	"math/rand"
+	"net"
+	"path/filepath"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/google/syzkaller/pkg/instance"
+	"github.com/google/syzkaller/pkg/mgrconfig"
+	"github.com/google/syzkaller/pkg/osutil"
+	"github.com/google/syzkaller/pkg/report"
+	"github.com/google/syzkaller/pkg/rpctype"
+	"github.com/google/syzkaller/pkg/tool"
+	"github.com/google/syzkaller/prog"
+	"github.com/google/syzkaller/syz-verifier/verf"
+	"github.com/google/syzkaller/vm"
+)
+
+// Verifier TODO
+type Verifier struct {
+	pools  map[int]*poolInfo
+	vmStop chan bool
+	// Location of a working directory for all VMs for the syz-verifier process.
+	// Outputs here include:
+	// - <workdir>/crashes/<OS-Arch>/*: crash output files grouped by OS/Arch
+	// - <workdir>/corpus.db: corpus with interesting programs
+	// - <workdir>/<OS-Arch>/instance-x: per VM instance temporary files
+	// grouped by OS/Arch
+	workdir     string
+	crashdir    string
+	target      *prog.Target
+	runnerBin   string
+	executorBin string
+	choiceTable *prog.ChoiceTable
+	rnd         *rand.Rand
+	addr        string
+}
+
+// RPCServer is a wrapper around the rpc.Server. It communicates with  Runners, generates programs and sends complete Results for verification.
+type RPCServer struct {
+	vrf     *Verifier
+	port    int
+	mu      sync.Mutex
+	progIdx int
+	pools   map[int]*poolInfo
+	progs   map[int]*programInfo
+}
+
+// poolInfo contains kernel-specific information for spawning virtual machines and reporting crashes. It also keeps track of the Runners executing on spawned VMs, what programs have been sent to each Runner and what programs have yet to be sent on any of the Runners.
+type poolInfo struct {
+	cfg      *mgrconfig.Config
+	pool     *vm.Pool
+	Reporter report.Reporter
+	//  vmRunners keep track of what programs have been sent to each Runner. There is one Runner executing per instance.
+	vmRunners map[int][]int
+	// progs stores the indices of programs that haven't been sent to this kernel yet but have been sent to at least one kernel.
+	progs []int
+}
+
+type programInfo struct {
+	prog       *prog.Prog
+	idx        int
+	serialized []byte
+	res        []*verf.Result
+	// left contains the indices of kernels that haven't sent results for this program yet.
+	left map[int]bool
```

Comment:
Holding both `res` and `left` here feels too complicated. We need to fill one initially, then add to one and remove from the other, etc.
The number of pools is a global constant, we know how many results we need. It feels it can be simpler: we want N results, so once we have N we are done.
Please think how to simplify this.

Reply:
I suspect it can be simpler if we don't special-case "VM has crashed", but instead consider it as also "receiving a result", just a somewhat special result (say, denoted as nil *verf.Result).

Reply:
But then the array containing results would have `nil` results as well and we'll need to ignore them in the `verf.Verify` function. Would that be better than using `left`?

Reply:
Good question.

Another possible option is to count both real results and crashes with a counter. Then when the counter reaches N, we know it's ready for verification, but `res` still contains only actual results.

It's somewhat similar to use of `left`. But I guess what bothers me about `left` is that it's a map, so it needs to be created, then filled and all that is additional lines of code, it also consumes memory and operations on it are slower than incrementing a counter. Also we count in both directions (res grows, while left shrinks), which feels slightly more complex to me than just counting in one direction.

But, yes, it's somewhat subjective. So maybe @melver can chime in?

I am also thinking how we will want to log the results. I think for each program we will want to log results from each kernel and in-order they are specified in the config (rather than in random order each time depending on how we received them). And for that it may be useful to store results in-order along the lines of:
```
p.res[pool.index] = res
p.numResults++
if p.numResults == numPools {
  verify(p)
}
```

Reply:
> Another possible option is to count both real results and crashes with a counter. Then when the counter reaches N, we know it's ready for verification, but res still contains only actual results.

I think that would be a good option as well, I agree that `left` makes the code more complex.

> Another possible option is to count both real results and crashes with a counter. Then when the counter reaches N, we know it's ready for verification, but res still contains only actual results.

I wasn't aware that we might want that. In that case your suggestion with `nil` results is probably the best.